presto-hive/src/main/java/com/facebook/presto/hive/util/StatisticsUtils.java

```diff
@@ -53,18 +57,35 @@ public static Partition incrementStatisticsCounters(Partition partition, HiveBas
         // whenever new data is inserted all statistics for that data must be collected
         checkAllStatisticsArePresent(statistics);
         Map<String, String> parameters = partition.getParameters();
-        Map<String, String> updatedParameters = incrementStatisticsCounters(parameters, statistics);
+        Map<String, String> updatedParameters = incrementStatisticsCounters(parameters, statistics, OptionalInt.empty());
         return Partition.builder(partition)
                 .setParameters(updatedParameters)
                 .build();
     }
 
-    private static Map<String, String> incrementStatisticsCounters(Map<String, String> parameters, HiveBasicStatistics statistics)
+    public static OptionalInt getPartitionCount(Map<String, String> parameters)
+    {
+        String value = parameters.get(NUM_PARTITIONS);
+        if (value == null) {
+            return OptionalInt.empty();
+        }
+        Integer numPartitions = Ints.tryParse(value);
+        if (numPartitions == null) {
+            return OptionalInt.empty();
+        }
+        return OptionalInt.of(numPartitions);
+    }
+
+    private static Map<String, String> incrementStatisticsCounters(Map<String, String> parameters, HiveBasicStatistics statistics, OptionalInt partitions)
     {
         HiveBasicStatistics currentStatistics = HiveBasicStatistics.createFromPartitionParameters(parameters);
         HiveBasicStatistics updatedStatistics = currentStatistics.add(statistics);
         Map<String, String> updatedParameters = new HashMap<>(parameters);
         updatedParameters.putAll(updatedStatistics.toPartitionParameters());
+        if (partitions.isPresent()) {
+            OptionalInt currentPartitionCount = getPartitionCount(parameters);
+            currentPartitionCount.ifPresent(count -> updatedParameters.put(NUM_PARTITIONS, Integer.toString(count + partitions.getAsInt())));
```

Comment:
https://cwiki.apache.org/confluence/display/Hive/StatsDev refers to "Fictional metastore property: `numPartitions`". Is this supposed to be a read-only table property?

Reply:
Yeah. But until we are on Hive 2.3+ we need to maintain a non-fictional property.

Reply:
Ok. And what happens in Hive 2.3+? Does it complain when you set this? (I know product tests use Hive >= 2, not sure which exactly)